### DISCUSSION TITLE
Child renders before the parent, is this an expected behaviour?

### DISCUSSION THREAD

**Discussion by emberist** [pmndrs/zustand#1477] [Q&A] [answered]
As the title says I'm having a issue using this approach:

```
export const useResourceStore = create((set, get) => ({
  resource: null,
  isFetching: false,
  fetch: async () => {
      if (get().resource || get().isFetching) {
         return;
      }

      set({ isFetching: true })

      // Fetching the resource
      const resource = ...

      set({ isFetching: false, resource })
  }
}))
```

Then I have a component that renders his children only if the resource have been fetched:

```
const ResourceFetcher = ({ children }) => {
    const { fetch, resource } = useResourceStore() 

    useEffect(() => {
       fetch()
    }, [fetch])

    if(!resource){
        return null
    }

    return <>{children}</>
}
```

Then Inside a child component of `ResourceFetcher` I use an hook like this one:

```
const useResource = () => {
    const resource = useResourceStore(s => s.resource);

    invariant(resource, "This hook is meant to be used inside a ResourceFetcher component")

    return resource
}
```

What is happening is that the invariant is failing, cause the Child component is rendered instead of `ResourceFetcher`, anyone knows why is this happening?

**Comment by dai-shi** [comment, marked as answer]
The first guess: Does it reproduce with React 18 too?

**Reply by emberist** [reply]
Mmh I'll try, I have to update the react dependency

**Reply by emberist** [reply]
Upgrading to react@18 have fixed the issue! Thanks

**Reply by dai-shi** [reply]
Thanks for confirming.
In this case, if someone needs a solution with react 17, `unstable_batchedUpdates` can help.
